WireDomain/Sources/WireDomain/Event Decryption/UpdateEventDecryptor.swift

```diff
@@ -0,0 +1,137 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireAPI
+import WireDataModel
+
+// sourcery: AutoMockable
+/// Decrypt the E2EE content within update events.
+protocol UpdateEventDecryptorProtocol {
+
+    /// Decrypt events in the given event envelope.
+    ///
+    /// - Parameter eventEnvelope: An event envelope that contains events received from the server.
+    /// - Returns: A list of decrypted update events.
+
+    func decryptEvents(in eventEnvelope: UpdateEventEnvelope) async throws -> [UpdateEvent]
+
+}
+
+struct UpdateEventDecryptor: UpdateEventDecryptorProtocol {
+
+    private let proteusService: any ProteusServiceInterface
+    private let context: NSManagedObjectContext
+
+    init(
+        proteusService: any ProteusServiceInterface,
+        context: NSManagedObjectContext
+    ) {
+        self.proteusService = proteusService
+        self.context = context
+    }
+
+    func decryptEvents(in eventEnvelope: UpdateEventEnvelope) async throws -> [UpdateEvent] {
+        let logAttributes: LogAttributes = [
+            .eventId: eventEnvelope.id.safeForLoggingDescription,
+            .public: true
+        ]
+
+        var decryptedEvents = [UpdateEvent]()
+
+        for event in eventEnvelope.events {
+            switch event {
+            case .conversation(.proteusMessageAdd(let eventData)):
+                WireLogger.updateEvent.info(
+                    "decrypting proteus event...",
+                    attributes: logAttributes
+                )
+
+                do {
+                    let decryptor = ProteusMessageDecryptor(
+                        proteusService: proteusService,
+                        managedObjectContext: context
+                    )
+
+                    let decryptedEventData = try await decryptor.decryptedEventData(from: eventData)
+                    decryptedEvents.append(.conversation(.proteusMessageAdd(decryptedEventData)))
+
+                } catch let error as ProteusError {
+                    WireLogger.updateEvent.error(
+                        "failed to decrypt proteus event payload, dropping: \(error.localizedDescription)",
+                        attributes: logAttributes
+                    )
+
+                    await appendFailedToDecryptProteusMessage(
+                        eventData: eventData,
+                        error: error
+                    )
+                } catch {
+                    WireLogger.updateEvent.error(
+                        "failed to decrypt proteus event, dropping: \(error.localizedDescription)",
+                        attributes: logAttributes
+                    )
```

Comment:
How do you mean?

Reply:
I mean in the case where `error` is a `ProteusError` we append it with `appendFailedToDecryptProteusMessage`. However we don't call `appendFailedToDecryptProteusMessage` in the default case, even though the error message includes `failed to decrypt proteus event`.

This was just a sanity check comment to make sure you didn't accidentally miss something, but please feel free to resolve this message.

Reply:
Ah I see now. I was preserving the existing behavior where we only appending decryption error system messages if it's a proteus error (because I believe the error is used to determine the message content).